Add testcases for feature related api's

VaibhavKansagara opened the pull request:
These include unit testcases for the 4 features such as TfFeature,IdfFeature, TfIdfDoclenFeature,TfDoclenCollTfCollLenFeature.Additionally it includes test
for FeatureList class which includes testing an empty featurelist and a
featurelist containing more than it's default features and also for generating
a database with three documents in which one has no common terms with other two.
Also consists of testing the output of create_feature_vetors method for all six
features like TfIdfDoclenFeature etc.

ojwb commented:
If I rebase onto current master and resolve the conflicts in what I believe is the correct way, I see 4 test failures:

```
$ ./runtest ./apitest -v checkcreatefeaturevector_idffeature checkcreatefeaturevector_tfidfdoclenfeature checkcreatefeaturevector_tfdoclencolllfcolllen checkemptyfeaturelist
Running test './apitest -v checkcreatefeaturevector_idffeature checkcreatefeaturevector_tfidfdoclenfeature checkcreatefeaturevector_tfdoclencolllfcolllen checkemptyfeaturelist' under eatmydata and valgrind
Running tests with backend "none"...
Running tests with backend "inmemory"...
Running tests with backend "glass"...
Running test: checkemptyfeaturelist...==31681== 24,116 (560 direct, 23,556 indirect) bytes in 2 blocks are definitely lost in loss record 33 of 35
==31681==    at 0x4835DEF: operator new(unsigned long) (vg_replace_malloc.c:334)
==31681==    by 0x48CDC73: Xapian::FeatureList::create_feature_vectors(Xapian::MSet const&, Xapian::Query const&, Xapian::Database const&) const (featurelist.cc:119)
==31681==    by 0x131A96: test_checkemptyfeaturelist() (api_letor.cc:301)
==31681==    by 0x1447CB: test_driver::runtest(test_desc const*) (testsuite.cc:335)
==31681==    by 0x14644C: test_driver::do_run_tests(__gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >) (testsuite.cc:644)
==31681==    by 0x14691E: run_tests (testsuite.cc:607)
==31681==    by 0x14691E: test_driver::run(test_desc const*) (testsuite.cc:829)
==31681==    by 0x125BEC: ApiTestRunner::run() const (api_collated.h:43)
==31681==    by 0x14363D: TestRunner::do_tests_for_backend(BackendManager*) (testrunner.cc:103)
==31681==    by 0x143873: TestRunner::run_tests(int, char**) (testrunner.cc:131)
==31681==    by 0x125590: main (apitest.cc:98)
==31681== 
==31681== LEAK SUMMARY:
==31681==    definitely lost: 560 bytes in 2 blocks
==31681==    indirectly lost: 23,556 bytes in 26 blocks
==31681==      possibly lost: 0 bytes in 0 blocks
==31681==    still reachable: 107,882 bytes in 32 blocks
==31681==         suppressed: 0 bytes in 0 blocks
==31681== Reachable blocks (those to which a pointer was found) are not shown.
==31681== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==31681== 
 LEAKED 24116 BYTES
Running test: checkcreatefeaturevector_idffeature... FAILED
api_letor.cc:1081: TEST_EQUAL_DOUBLE_((fvals_doc1[0]), (0.0))
Expected 'fvals_doc1[0]' and '0.0' to be (nearly) equal: were 1 and 0)


Running test: checkcreatefeaturevector_tfidfdoclenfeature... FAILED
api_letor.cc:1307: TEST_EQUAL_DOUBLE_((fvals_doc1[0]), (0.0))
Expected 'fvals_doc1[0]' and '0.0' to be (nearly) equal: were 1 and 0)


Running test: checkcreatefeaturevector_tfdoclencolllfcolllen... FAILED
api_letor.cc:1388: TEST_EQUAL_DOUBLE_((fvals_doc1[1]), (test_vals_doc1[1] / max_val[1]))
Expected 'fvals_doc1[1]' and 'test_vals_doc1[1] / max_val[1]' to be (nearly) equal: were 0.979504930813703 and 0.979061300626175)


./apitest backend glass: 0 tests passed, 4 failed.
Running tests with backend "singlefile_glass"...
Running tests with backend "multi_glass"...
Running tests with backend "remoteprog_glass"...
Running tests with backend "remotetcp_glass"...
./apitest total: 0 tests passed, 4 failed.
```

The memory leak certainly seems genuine - if there are no features then we still create a new `Feature::Internal` but we never assign it anywhere so the reference counting doesn't get a chance to delete it again.

I think the check for `mset.empty()` at the top of `FeatureList::create_feature_vectors()` should also check for `internal->feature.empty()` - there's no point iterating the MSet if there are no features.

I didn't look at the others.

Let me know if you want me to push my rebase branch somewhere public.

ojwb commented:
I pushed it here, so it's available right away if it's useful:

https://github.com/ojwb/xapian/tree/unittestcases-rebased

VaibhavKansagara commented:
> I pushed it here, so it's available right away if it's useful:
> 
> https://github.com/ojwb/xapian/tree/unittestcases-rebased

thanks

VaibhavKansagara commented:
> I think the check for mset.empty() at the top of FeatureList::create_feature_vectors() should also check for internal->feature.empty() - there's no point iterating the MSet if there are no features.
>
Weight will be the default feature when there are no features so I think we do need to iterate through the msets. I have fixed this in #264 and i have also fixed the testcase failures in this PR.

jaylett commented on xapian-letor/tests/api_letor.cc:
```diff
@@ -229,6 +280,36 @@ DEFINE_TESTCASE(emptyfeaturelist, !backend)
+    f.push_back(new Xapian::TfIdfDoclenFeature());
+    f.push_back(new Xapian::TfDoclenCollTfCollLenFeature());
+    f.push_back(new Xapian::TfFeature());
+    f.push_back(new Xapian::TfDoclenFeature());
```
To avoid people having to check the feature implementations in future, it might be worth having comments saying how many values each feature contributes to the feature vector, which would explain where the number 25 comes from later in the testcase.

VaibhavKansagara replied:
fixed in a fixup commit.

jaylett commented on xapian-letor/tests/api_letor.cc:
```diff
@@ -229,6 +280,36 @@ DEFINE_TESTCASE(emptyfeaturelist, !backend)
+    Xapian::MSet mset;
+    mset = enquire.get_mset(0, 10);
+
+    TEST(!mset.empty());
```
I think we should either not test this at all (since this is core functionality, not letor), or test explicitly that the MSet size is 2 — since we test later that we make 2 feature vectors, which will be true if and only if there are two items in the MSet. (Since we only have two documents in the db, this is reasonably implicit, so I'd be happy just to drop this `TEST()`.)

VaibhavKansagara replied:
fixed in a fixup commit.

jaylett commented on xapian-letor/tests/api_letor.cc:
```diff
@@ -833,3 +914,456 @@ DEFINE_TESTCASE(ndcg_score_test, generated)
+    queryparser.add_prefix("description", "XD");
+
+    string querystring = "title:score description:score score";
+    Xapian::Query query = queryparser.parse_query(querystring);
```
This needs a comment explaining why we're doing this rather than just using `Query("score")`.

VaibhavKansagara replied:
fixed in a fixup commit.

jaylett commented on xapian-letor/tests/api_letor.cc:
```diff
@@ -833,3 +914,456 @@ DEFINE_TESTCASE(ndcg_score_test, generated)
+    vector<double> test_vals_doc1(4);
+    test_vals_doc1[0] = 0.301029995663981;
+    test_vals_doc1[1] = 1.653212513775344;
+    test_vals_doc1[2] = 1.954242509439325;
```
I think we need a brief comment explaining where these numbers come from. (Even though the answer is just that they are the appropriate TF values for the first document.)

VaibhavKansagara replied:
fixed in a fixup commit.

jaylett commented on xapian-letor/tests/api_letor.cc:
```diff
@@ -1006,14 +1010,14 @@ DEFINE_TESTCASE(createfeaturevector_idffeature, generated)
     queryparser.add_prefix("description", "XD");
 
+    // As the feature values depend on "title","body" and "whole",
+    // we need to seperate it out instead of just writing Query("score").
```
I still had to check the `TfFeature` implementation. Can we maybe say "as the feature values depend on different prefixed terms in the query" or something like that?

Also, it's "separate" not "seperate" (which is causing the CI failure, since we have a check for that in `xapian-check-patch` — it's a good habit to run this yourself every so often because a CI run takes a fair amount of time, so fixing these problems before you push makes everything smoother).

VaibhavKansagara replied:
I did run it before pushing it but I failed to realize I was running it before pushing the changes and hence it was showing the results of the previous run and so I was not able to catch it.
Fixed in a fixup commit.

ojwb replied:
It just takes a diff as input, so you can get it to check uncommitted changes like so:

```
git diff|xapian-maintainer-tools/xapian-check-patch
```

Or `git diff --cached|...` if you want to check changes staged for commit, or `git show SHA|...` if you want to check the changes in the single commit with sha `SHA`, and so on.

ojwb commented:
i don't have any further comments, but I should probably let @jaylett check the latest changes before merging since they were in response to his comments.

jaylett commented:
LGTM 👍

ojwb commented:
Thanks - rebased, autosquashed and pushed as c4b7eb5222dbc2cc43410e5c64b97a10c20e3985.
